planning/behavior_path_planner/autoware_behavior_path_lane_change_module/src/utils/calculation.cpp

```diff
@@ -382,30 +382,58 @@ std::vector<double> calc_lon_acceleration_samples(
 }
 
 double calc_lane_changing_acceleration(
-  const double initial_lane_changing_velocity, const double max_path_velocity,
-  const double lane_changing_time, const double prepare_longitudinal_acc)
+  const CommonDataPtr & common_data_ptr, const double initial_lane_changing_velocity,
+  const double max_path_velocity, const double lane_changing_time,
+  const double prepare_longitudinal_acc)
 {
   if (prepare_longitudinal_acc <= 0.0) {
-    return 0.0;
+    const auto & params = common_data_ptr->lc_param_ptr->trajectory;
+    const auto lane_changing_acc = common_data_ptr->transient_data.is_ego_near_current_terminal_start
+                                    ? prepare_longitudinal_acc * params.lane_changing_decel_factor
+                                    : 0.0;
+    return lane_changing_acc;
   }
 
   return std::clamp(
     (max_path_velocity - initial_lane_changing_velocity) / lane_changing_time, 0.0,
     prepare_longitudinal_acc);
 }
 
+double calc_actual_prepare_duration(
+  const CommonDataPtr & common_data_ptr, const double current_velocity,
+  const double active_signal_duration)
+{
+  const auto & params = common_data_ptr->lc_param_ptr->trajectory;
+  const auto min_lc_velocity = params.min_lane_changing_velocity;
+
+  // need to ensure min prep duration is sufficient to reach minimum lane changing velocity
+  const auto min_prepare_duration = std::invoke([&]() -> double {
+    if (current_velocity >= min_lc_velocity) {
+      return params.min_prepare_duration;
+    }
+    const auto max_acc =
+      std::min(common_data_ptr->bpp_param_ptr->max_acc, params.max_longitudinal_acc);
+    if (max_acc < eps) {
+      return params.max_prepare_duration;
+    }
+    return (min_lc_velocity - current_velocity) / max_acc;
```

Comment:
since we ensure `min_lane_changing_velocity` is feasible [here](https://github.com/autowarefoundation/autoware.universe/pull/9185/files#diff-fec5205c985f56e5b8fac2fbad0352bab1edfe1542ee916da103e94dcf60bbddR71-R72), even if `current_velocity` is 0.0, the prepare duration should not exceed `max_prepare_duration`